### ISSUE TITLE
Configurable keyboard shortcuts

### ISSUE THREAD

**Issue by jameskerr** [jameskerr/react-arborist#57]
All the keyboard shortcuts are listed in the DefaultContainer component. I could see how people would want to use their own keybindings.

Something like:

```js
{
  "a": "createLeaf",
  "shift+a": "createInternal"
  "space": "activate",
  "meta+DownArrow": "activate"
}
```

**Comment by amarnath-cohesive** [comment]
Yes! I wish there was a way to do this. Also, I am trying to bind these functions to some buttons outside of the tree. I tried creating a ref and passing it to the Tree component, and the using the ref to do `tree.createLeaf()`. But it doesn't seem to be working.

**Comment by jameskerr** [comment]
Can you post the code that's not working? 

If it's a ref, you may need to `tree.current.createLeaf()`. 

Also, if you're using the `data` prop. you'll need to have an `onCreate` prop on the tree in order for `createLeaf()` work.

**Comment by amarnath-cohesive** [comment]
I should have used `tree.current.createLeaf()`. Thanks for the response.

**Comment by hilmanauz** [comment]
hi @jameskerr, is there any way to disable the keyboard shortcuts from the Container props? I want to disable space shortcuts to toggle the parent folder

**Comment by jameskerr** [comment]
@hilmanauz You can always intercept the onKeyDown event in your RowRenderer and call `e.stopPropagation()`. That's a workaround for today.

**Comment by jameskerr** [comment]
@holloway I want to break up that giant onKeyDown event listener found in DefaultContainer. Each of those cases should be a named function with the TreeApi as the only argument. All of those could go in their own file called commands.ts.

The commands should be put in an object; the key is the name, the value is function.

Another object will map a keybinding string to the string name of the command function. A user could then configure this by passing their own object.

Then in the DefaultContainer#onKeyDown handler, we have some logic that turns that event into a keybinding string. We could follow VSCode's example for how to structure the string (`Shift+ArrowDown`) for example. Then we run whatever command is mapped to that keybinding string using the object above.

I'm wondering if we want to only make a few of the keybindings available for customization. For example, "ArrowDown" should probably not be configurable. But "Space" should be.

Something Like

```tsx
<DefaultContainer
  onKeyDown={(e) => {
    const keybinding = toKeyBinding(e)
    const command = tree.runCommandAt(keybinding)
  }}
/>

class TreeApi() {
  runCommand(key: string) {
    const commandName = this.keybindings[key]
    const command = this.commands[commandName]
    if (command) command(this)
  }
}

```

**Comment by amarnath-cohesive** [comment]
@hilmanauz , what I did to solve this was to host the library on my own github after removing the onKeyDown listener from the DefaultContainer file.

**Comment by hilmanauz** [comment]
it works for me @jameskerr, thanks for your advice. I forgot there's an onKeyDown prop in every element.

**Comment by holloway** [comment]
Hi @jameskerr see #76 ... keen for a review :smile:

**Comment by stefantrinh1** [comment]
> @hilmanauz You can always intercept the onKeyDown event in your RowRenderer and call `e.stopPropagation()`. That's a workaround for today.

Hey is anyone able to show an example of this at work. 

I tried this on the tree and the custom node component

```
           <Tree
              idAccessor="_id"
              data={tree}
              width={editorWidth}
              height={height - 132 ?? 500}
              rowHeight={78}
              openByDefault={false}
              onCreate={onCreate}
              onRename={onRename}
              onMove={onMove}
              onDelete={onDelete}
              onKeyDown={e => {
                e.stopPropagation()
              }}
            >
              {props => {
                return <Node   onKeyDown={e => {
                  e.stopPropagation()
                }} setTree={setTree} {...props} />
              }}
            </Tree>
 ```

**Comment by jameskerr** [comment]
Work on this is progressing. Today I just finished it in the use-nodes branch. 

Keyboard shortcuts look like this:

```js
import { ShortcutAttrs } from "./types";

export const defaultShortcuts: ShortcutAttrs[] = [
  /* Keyboard Navigation */
  { key: "ArrowDown", command: "focusNext" },
  { key: "ArrowUp", command: "focusPrev" },
  { key: "ArrowLeft", command: "focusParent", when: "isLeaf || isClosed" },
  { key: "ArrowLeft", command: "close", when: "isOpen" },
  { key: "ArrowRight", command: "open", when: "isClosed" },
  { key: "ArrowRight", command: "focusNext", when: "isOpen" },
  { key: "Home", command: "focusFirst" },
  { key: "End", command: "focusLast" },
  { key: "PageDown", command: "focusNextPage" },
  { key: "PageUp", command: "focusPrevPage" },

  /* Tabbing Around */
  { key: "Tab", command: "focusOutsideNext" },
  { key: "Shift+Tab", command: "focusOutsidePrev" },

  /* CRUD */
  { key: "Backspace", command: "destroy" },
  { key: "a", command: "createLeaf" },
  { key: "Shift+A", command: "createInternal" },
  { key: "Enter", command: "edit" },

  /* Selection */
  { key: "Shift+ArrowUp", command: "moveSelectionStart" },
  { key: "Shift+ArrowDown", command: "moveSelectionEnd" },
  { key: " ", command: "toggle", when: "isInternal" },
  { key: " ", command: "select", when: "isLeaf" },
  { key: "Meta+a", command: "selectAll" },
  { key: "Control+a", command: "selectAll" },
];
```

**Comment by jessicarush** [comment]
Thank you for all the work you're doing, I'm really looking forward to this feature. 

Just one thought: since a mouse click will "select" when clicking a leaf node and both "toggle" and "select" when clicking a node that has children, it may make sense to have the spacebar key behave the same way in that it both selects and toggles an `isInternal`.

**Comment by sauron918** [comment]
People do really want to use their own keybindings :) Thanks!

**Comment by jameskerr** [comment]
Thanks for the comments and suggestions. I'm excited to ship more of these features, but I'm busy with other projects at the moment. Everything takes longer than first expected!

**Comment by TrevorBurnham** [comment]
Using this as the tracking issue for keyboard customization. Related requests rolled up here:

- #268 Custom action on keyDown (override Enter)
- #72 Keyboard navigation breaks when pressing Enter on a node without a text input
- #36 Ctrl+Z (undo)
